logprep/util/auto_rule_corpus_tester.py

```diff
@@ -0,0 +1,483 @@
+        data_directory = self.input_test_data_path
+        file_paths = self._collect_test_case_file_paths(data_directory)
+        test_cases = self._group_path_by_test_case(data_directory, file_paths)
+        self.test_cases = dict(sorted(test_cases.items()))
+
+    def _group_path_by_test_case(self, data_directory, file_paths):
+        test_cases = {}
+        for filename in file_paths:
+            test_case_id = self._strip_input_file_type(filename)
+            if test_case_id not in test_cases:
+                test_cases[test_case_id] = {
+                    "test_data_path": {"in": "", "out": "", "out_extra": ""},
+                    "report_print_statements": [],
+                }
+            data_path = test_cases.get(test_case_id, {}).get("test_data_path", {})
+            if "_in.json" in filename:
+                data_path.update({"in": os.path.join(data_directory, filename)})
+            if "_out.json" in filename:
+                data_path.update({"out": os.path.join(data_directory, filename)})
+            if "_out_extra.json" in filename:
+                data_path.update({"out_extra": os.path.join(data_directory, filename)})
+            test_cases[test_case_id]["test_data_path"].update(data_path)
+        return test_cases
+
+    def _collect_test_case_file_paths(self, data_directory):
+        file_paths = []
+        for root, _, files in os.walk(data_directory):
+            for filename in files:
+                file_paths.append(os.path.abspath(os.path.join(root, filename)))
+        return file_paths
+
+    def _run_logprep_per_test_case(self):
+        """
+        For each test case the logprep connector files are rewritten (only the current test case
+        will be added to the input file), the pipline is run and the outputs are compared.
+        """
+        for current_test_case in self.test_cases.items():
+            if not current_test_case[1].get("test_data_path", {}).get("in"):
+                raise ValueError(
+                    f"The test case '{current_test_case[0]}' is missing an input file."
+                )
+            self._prepare_connector_files(current_test_case)
+            self.pipeline._process_pipeline()
+            output = self._retrieve_pipeline_output()
+            current_test_case[1].update({"logprep_output": output})
+
+    def _compare_with_expected_outputs(self):
+        """Compare the generated logprep output with the current test case"""
+        self.console.print("[b]# Test Cases Summary:")
+        for test_case_id in self.test_cases:
+            self._compare_and_collect_report_print_statements(test_case_id)
+            self._print_short_test_result(test_case_id)
+            if len(self.test_cases.get(test_case_id, {}).get("report_print_statements", [])) > 0:
+                self.at_least_one_failed = True
+
+    def _compare_and_collect_report_print_statements(self, test_case_id):
```

Comment:
a method called ...AND... gives me not a good feeling ;)
Consider to refactor it in two methods

Reply:
I have revised most of the methods and renamed them also if necessary.